### ISSUE TITLE
 LDAP disables hostname verification

### ISSUE THREAD

**Issue by nmco** [geosolutions-it/geoserver#204]
See:

https://lgtm.com/projects/g/geoserver/geoserver/snapshot/ee2246871b8da41516c7a8154c2d3ecd69b34570/files/src/security/ldap/src/main/java/org/geoserver/security/ldap/LDAPUtils.java?sort=name&dir=ASC&mode=heatmap#L46

Should probably be made configurable, seems something that would be useful in dev/test/integration environments.

Related security ticket is available [here](https://github.com/geosolutions-it/geoserver-security/issues/28).

**Comment by nmco** [comment]
@alemarte let's proceed by:
1. Investigate the issue, and reproduce it.
1. Propose a plan and estimate to fix it, you should take into account that: 
   1. Looks like the direction we want to go is that the host-name verification should be always active by default but it should be possible to deactivate it to perform some tests.
   1. Having this verification deactivated can lead to a security breach, hence some kind of warning would need to be displayed on GeoServer main page.
1. Let's discuss your plan and estimate, wait for my green light before moving on whit the implementation.

**Comment by alemarte** [comment]
Unfortunately I haven't been able to reproduce the issue yet. For what I've seen setting the TLS check to true automatically disable the TLS hostname verification check. My suggestion is to introduce a second checkbox once the first is enabled, which specifies whether the default hostname verification check should be disabled (default no, highlighting the associated risk with a warning).

**Comment by nmco** [comment]
Meeting with 18-09-2020 minutes:
- @alemarte follow the training material guide for LDAP but it didn't work, although looks like an old version of the virtual VM was used.
- @alemarte will give it another try following [the training for LDAP](https://docs.geoserver.geo-solutions.it/edu/en/security/ldap_authentication.html) and used the [latest GeoServer training package](https://www.dropbox.com/s/hg13fjzrvdtgoss/Training-Geoserver-2.17.x-V3.zip?dl=1).

**Comment by alemarte** [comment]
@nmco 
Plan:

- Introduce a second checkbox **hostnameVerification** which appears once the first checkbox "useTLS" is checked.
- The **hostnameVerification** is disabled by default.
- If the **hostnameVerification** is checked then the always true HostnameVerifier is configured into the LDAP context.
- Once the **hostnameVerification** is checked and the LDAP filter is enabled, a warning appears on the geoserver homepage.

About test
- I would test the following cases:

1.  The "useTLS" is true and **hostnameVerification** is false 
     => The LdapContext has no HostnameVerifier configured (so that default check is enabled)
2. The "useTLS is true and **hostnameVerification** is true 
    => The LdapContext has the custom HostnameVerifier (I could implement a named HostnameVerified instead of the Anonymous one and check the type).

Estimate 8h

**Comment by nmco** [comment]
Thx for the plan @alemarte, I have a few comments:

> Introduce a second checkbox hostnameVerification which appears once the first checkbox "useTLS" is checked.

I don't know much about LDAP, so just double checking, we can only enabled host name verification if [TLS is used](https://docs.geoserver.org/stable/en/user/security/auth/providers.html#secure-ldap-connections)? 

> The hostnameVerification is disabled by default.

I was expecting the opposite, by default the host name verification should be activated, since this is what offers a secure environment. It will only be deactivated explicitly in special conditions.

 The rest looks fine to me.

**Comment by alemarte** [comment]
You're right. The **hostnameVerification** checkbox should be named **disableHostnameVerification**. And the default underlying behaviour is to always check.

**Comment by nmco** [comment]
Great, go ahead you have my green light :+1:.

**Comment by alemarte** [comment]
JIRA Ticket https://osgeo-org.atlassian.net/browse/GEOS-9742

**Comment by alemarte** [comment]
@aaime @nmco 
An LDAP configuration parameter has been added [tlsHostnameVerificationDisabled ](https://github.com/alemarte/geoserver/blob/f0836725b119a327d8f3a60f655ccf7dea6cf762/src/security/ldap/src/main/java/org/geoserver/security/ldap/LDAPBaseSecurityServiceConfig.java#L35).

Now if this parameter is true it must be displayed in the home page as a security warning.

Basically in the [SecurityHomePageContentProvider ](https://github.com/alemarte/geoserver/blob/f0836725b119a327d8f3a60f655ccf7dea6cf762/src/web/security/core/src/main/java/org/geoserver/security/web/SecurityHomePageContentProvider.java#L193) the security manager should expose an abstract way to fetch that information, which in the main module isn't available.

I suggest to introduce an Interface GeoServerAuthenticationProviderSecurityWarning  with expose security warnings which each GeoServerAuthenticationProvider can decide to implement just in case.

Does it make sense?

**Comment by nmco** [comment]
LDAP is an extension, hence we can't make ``[SecurityHomePageContentProvider](https://github.com/geoserver/geoserver/blob/8672a10413166c553eb7f4fcf0d2a250337561a0/src/web/security/core/src/main/java/org/geoserver/security/web/SecurityHomePageContentProvider.java#L35)`` aware of it, rephrasing we can't create a dependency between GeoServer core and an extension.

In my opinion, the best way to go here is to create an extension point in ``[SecurityHomePageContentProvider](https://github.com/geoserver/geoserver/blob/8672a10413166c553eb7f4fcf0d2a250337561a0/src/web/security/core/src/main/java/org/geoserver/security/web/SecurityHomePageContentProvider.java#L35)`` that will allow other modules to contribute their own warnings.

I'm providing bellow pointers that shows how we can create such an extension point using Apache Wicket and Spring beans, I'm using as an example the services configuration pages that allow extension to contribute their own bit of configuration:
- The extension point is declared in HTML https://github.com/geoserver/geoserver/blob/4f5b375df95a121b0ac0a6bebd5b53efb9466ecc/src/web/core/src/main/java/org/geoserver/web/services/BaseServiceAdminPage.html#L63-L65.
- [Here](https://github.com/geoserver/geoserver/blob/1960e66c311f8de15d817498b2258eae1474234c/src/web/core/src/main/java/org/geoserver/web/services/BaseServiceAdminPage.java#L139-L142) and [here](https://github.com/geoserver/geoserver/blob/1960e66c311f8de15d817498b2258eae1474234c/src/web/core/src/main/java/org/geoserver/web/services/BaseServiceAdminPage.java#L204-L236) the beans are retrieved from Spring and then injected.

You can find an usage of the extension point:
- [Here](https://github.com/geoserver/geoserver/blob/6bb431a36a57aa1b6cf68d78bc7dea35b0958812/src/extension/inspire/src/main/java/org/geoserver/inspire/web/InspireAdminPanel.java#L41) is the implementation of the extension point bean.
- [Here](https://github.com/geoserver/geoserver/blob/4d057f24542e3680315809497a8c4435e9d2e377/src/extension/inspire/src/main/resources/applicationContext.xml#L23-L29) the bean is registered in spring.

In this case I would create an extension point that would allow  extension to contribute their label message, similar to [this](https://github.com/geoserver/geoserver/blob/8672a10413166c553eb7f4fcf0d2a250337561a0/src/web/security/core/src/main/java/org/geoserver/security/web/SecurityHomePageContentProvider.java#L154-L175).

Let me know if this unblock you @alemarte :small_airplane:.

**Comment by nmco** [comment]
After a quick chat with @aaime, there is two comments for you @alemarte:
- This page ``GeoServerHomePageContentProvider`` may already provide the extension point you need.
- This warning should only be displayed when the administrator is log in.

**Comment by nmco** [comment]
Work done by @alemarte is available here: https://github.com/alemarte/geoserver/tree/GEOS-9742.

**Comment by imranrajjad** [comment]
@nmco I am not very proficient with LDAP services, I will excuse myself from this

**Comment by zvrablik** [comment]
@aaime @nmco 
The acme ldap does it have any option to work with the TLS too? Could you point me to any documentation how to setup?
Only port I see opened when I start acme ldap is 10389. Connecting to this without tls works.

Docs I was lookig at
 to setup ldpa without lts https://docs.geoserver.org/stable/en/user/security/auth/providers.html#secure-ldap-connections
edu module https://docs.geoserver.geo-solutions.it/edu/en/security/ldap_authentication.html

Thanks.

**Comment by zvrablik** [comment]
I am able to make ldap working with the acme mock jar ldap. Works fine following steps [here](https://docs.geoserver.geo-solutions.it/edu/en/security/ldap_authentication.html)  

I can't use the acme jar file to test with TLS configuration I don't see open port 10636 so I think it does not have tls mock support.

I tried using OpenDJ, but I can't get it working even without TLS. I have created ldif file along description of the users/groups tree

@nmco  Whom could I talk to about alternative ldap server configuration. I tried use the [openDJ](https://github.com/OpenRock/OpenDJ/releases) as it has TLS support, but no luck to configure this with geoserver. I was using this ldap configuration [here](https://gist.github.com/391aa2d695d7422bdf7a20295bf2cbea) and my geoserver config is 
![geoserver config screenshot](https://user-images.githubusercontent.com/1183452/116203807-620fa180-a73c-11eb-8e95-cbc1544f3bdc.png)

**Comment by nmco** [comment]
Notes:
- @zvrablik uses linux, hence can't use directly the training package

@zvrablik let's have a quick session, I'm curious about what GeoServer uses for the integration tests.

**Comment by nmco** [comment]
@fernandor777 support @zvrablik installing and configuring Apache DS. @zvrablik still needs to validate that it works.

**Comment by zvrablik** [comment]
ApacheDS is now finally working. I was able to import there my ldif file containing acme structer as in Geoserver ldap documentation.
I still can't connect from geoserver, not sure why. I have verified in the ApacheStudio the user passwords, all looks good ...
TODO
* get working ldap with geoserver without TLS configuration
* get working ldap with geoserver and TLS configuration
* finish estimates

**Comment by nmco** [comment]
@zvrablik is unlocked, he was able to successfully start the LDAP. He is now investigating the issues.

**Comment by nmco** [comment]
@zvrablik let's time box the remaining investigation to 2 hour.

**Comment by zvrablik** [comment]
@aaime 

hi Andrea,
I have got issue configuring ldap. Would you have time to help me with it.

 I run into many issues we fixed with Fernando. Now I use Apahce DS as in our tests.
I use the data.ldif from our automated tests - imported into ApacheDS.

I think my setup should be now correct and running ldap test connection without ldaps works fine with same user.

What I have done:
I have generated new keystore using my machine name w530 . Added the keystore to ApacheDS using ApacheDS Studion.

I have exported certificate and added it to the java cacert
Geoserver does not complain about missing certificate , so I think this is correct.
but when I try the connection in LDAP Authentication Provider page I get this exception in logs

Caused by: org.springframework.ldap.UncategorizedLdapException: Uncategorized exception occured during LDAP processing; nested exception is javax.naming.NamingException: LDAP connection has 
been closed
        at org.springframework.ldap.support.LdapUtils.convertLdapException(LdapUtils.java:228)
        at org.springframework.ldap.core.support.AbstractContextSource.doGetContext(AbstractContextSource.java:147)
        at org.springframework.ldap.core.support.AbstractContextSource.getContext(AbstractContextSource.java:130)
        at org.springframework.security.ldap.authentication.BindAuthenticator.bindWithDn(BindAuthenticator.java:124)
        at org.springframework.security.ldap.authentication.BindAuthenticator.bindWithDn(BindAuthenticator.java:110)
        at org.springframework.security.ldap.authentication.BindAuthenticator.authenticate(BindAuthenticator.java:85)
        at org.geoserver.security.ldap.GeoserverLdapBindAuthenticator.authenticate(GeoserverLdapBindAuthenticator.java:54)
        at org.springframework.security.ldap.authentication.LdapAuthenticationProvider.doAuthentication(LdapAuthenticationProvider.java:187)
        ... 142 more
Caused by: javax.naming.NamingException: LDAP connection has been closed
        at java.naming/com.sun.jndi.ldap.LdapRequest.getReplyBer(LdapRequest.java:133)
        at java.naming/com.sun.jndi.ldap.Connection.readReply(Connection.java:443)
        at java.naming/com.sun.jndi.ldap.LdapClient.extendedOp(LdapClient.java:1206)
        at java.naming/com.sun.jndi.ldap.LdapCtx.extendedOperation(LdapCtx.java:3379)
        at java.naming/javax.naming.ldap.InitialLdapContext.extendedOperation(InitialLdapContext.java:184)
        at org.springframework.ldap.core.support.AbstractTlsDirContextAuthenticationStrategy.processContextAfterCreation(AbstractTlsDirContextAuthenticationStrategy.java:131)
        at org.springframework.ldap.core.support.AbstractContextSource.doGetContext(AbstractContextSource.java:142)
        ... 148 more


or  021-04-30 16:16:57,588 DEBUG [org.geoserver.security.filter.GeoServerUserNamePasswordAuthenticationFilter$1] - Request is to process authentication
2021-04-30 16:16:57,633 DEBUG [org.geoserver.security] - Bad credentials
org.springframework.security.authentication.BadCredentialsException: Bad credentials
        at org.springframework.security.authentication.dao.DaoAuthenticationProvider.additionalAuthenticationChecks(DaoAuthenticationProvider.java:93)
        at org.springframework.security.authentication.dao.AbstractUserDetailsAuthenticationProvider.authenticate(AbstractUserDetailsAuthenticationProvider.java:166)
        at org.geoserver.security.auth.UsernamePasswordAuthenticationProvider.authenticate(UsernamePasswordAuthenticationProvider.java:78)
        at org.geoserver.security.GeoServerAuthe

Would you have time to help me with this?

**Comment by aaime** [comment]
@zvrablik I have been busy in meetings all afternoon, and will be till the end of the day. But it does not matter much, I don't know how to use ApacheDS, the little work I did on LDAP I did using the Acme jar... :-(
Believe Fernando is the one among us that knows ApacheDS the best.

**Comment by aaime** [comment]
Yeah, looking at the module history, the commits belonging to me are GeoServer-wide passes of cleanups, often done using IntelliJ automatic cleanup procedures:

![image](https://user-images.githubusercontent.com/325433/116724141-95725a80-a9e0-11eb-9dce-cf0da7bba01d.png)

Think I tried to do something on it, and then the task was given to Fernando.

**Comment by zvrablik** [comment]
Thanks Andrea.

Current status:

I have created small cli program to test the ldaps communication and it worked : 



```
import java.util.Hashtable;
import javax.naming.*;
import javax.naming.directory.*;

public class ConnectWithLdaps {
        
    public static void main(String[] args) throws NamingException {
                
        Hashtable env = new Hashtable();
                
        // Simple bind
        env.put(Context.SECURITY_AUTHENTICATION, "simple");
        env.put(Context.SECURITY_PRINCIPAL,
                "uid=admin,ou=People,dc=example,dc=com");
        env.put(Context.SECURITY_CREDENTIALS, "admin");
                
        env.put(Context.INITIAL_CONTEXT_FACTORY,
            "com.sun.jndi.ldap.LdapCtxFactory");
        env.put(Context.PROVIDER_URL, "ldaps://w530:10636/dc=example,dc=com");
                
        DirContext ctx = new InitialDirContext(env);
        NamingEnumeration enm = ctx.list("");
                
        while (enm.hasMore()) {
            System.out.println(enm.next());
        }
                
        enm.close();
        ctx.close();
    }
}
```

* I have generated private key for my machine using my machine name and imported it into ApacheDS
* I have imported the certificate of private key from keystore to java cacerts keystore 
* ApacheDS Studio is now able to connect using ldaps connection without no certificate found warning
* the code above works fine when I use java with updated cacets

* testing in geoserver but it does not work when I use url ldaps://w530:10636/dc=example,dc=com , request to ApacheDS never finish...

**Comment by zvrablik** [comment]
I have tested this:

enabled the TLS checkbox in geoserver and left the url as used in unsecured ldap e.g. ldap://w530:10389/dc=example,dc=com

I was debugging the place where ldaps got stuck and there was this

the request id was 1.3.6.1.4.1.1466.20037  that is request definition to start using TLS over standard ldap.

Looks like I was all time combining startTLS and ldaps and that cause the issue.

@nmco @fernandor777 
Looks like in Geoserver only supported secured connection is startTLS and ldaps is not working.
I have tested the ldaps ..., do we need working ldaps too?

**Comment by nmco** [comment]
@fernandor777 is busy for the time being, let's put this issue on hold.
